### ISSUE TITLE
Decide on wanted behavior for request+exit or exit+request together

### ISSUE THREAD

**Issue by foolip** [whatwg/fullscreen#63]
Working on Fullscreen in Blink, the situation has come up where we're transition in or out of fullscreen when a call to go in the opposite direction happens. Test case:

```HTML
<!doctype html>
<div>
 <button>request</button>
 <button>exit</button>
 <button>request+exit</button>
 <button>exit+request</button>
</div>
</div>
<script>
var div = document.querySelector('div');
div.addEventListener('click', function(event) {
  if (event.target.localName != 'button')
    return;
  for (var action of event.target.textContent.split('+')) {
    if (action == 'request') {
      if (div.requestFullscreen) {
        div.requestFullscreen();
      } else if (div.webkitRequestFullscreen) {
        div.webkitRequestFullscreen();
      } else {
        div.mozRequestFullScreen();
      }
    } else if (action == 'exit') {
      if (document.exitFullscreen) {
        document.exitFullscreen();
      } else if (document.webkitExitFullscreen) {
        document.webkitExitFullscreen();
      } else {
        document.mozCancelFullScreen();
      }
    } else {
      throw 'unknown action';
    }
  }
});
</script>
```

Current behavior:
<table>
<tr>
 <th></th>
 <th>request+exit</th>
 <th>exit+request</th>
 <th>request then request+exit</th>
 <th>request then exit+request</th>
<tr>
 <th>Spec'd</th>
 <th>enter</th>
 <th>enter</th>
 <th>???</th>
 <th>???</th>
<tr>
 <th>Chrome 55</th>
 <th>enters+exits quickly</th>
 <th>enters</th>
 <th>exits</th>
 <th>exits</th>
<tr>
 <th>Edge 14</th>
 <th>enters+exits quickly</th>
 <th>enters</th>
 <th>exits</th>
 <th>exits+enters quickly</th>
<tr>
 <th>Firefox 51</th>
 <th>enters</th>
 <th>enters</th>
 <th>exits</th>
 <th>exits</th>
</table>

Safari 10 behaved strangely when testing in BrowserStack, seemingly hanging, so I don't trust what I saw.

[Exit fullscreen](https://fullscreen.spec.whatwg.org/#exit-fullscreen) has an early return in the spec, which makes the first two cases simple. For the second two, it's not obvious well defined, because two "resizes" are racing "in parallel."

@upsuper @jernoble

**Comment by foolip** [comment]
@aliams too

**Comment by foolip** [comment]
Possible models:
 1. All `requestFullscren()` and `exitFullscreen()` calls go on a single cue, and are processed in order. (Would match Edge in the simple cases tested.)
 1. Keep track of pending resizes and reject any attempt to go in the other direction until done. (Would match Firefox in the simple cases tested.)
 1. In `requestFullscreen()`, attempt no resize if pending's top-level browsing context’s active document’s fullscreen element is not null, so that if already in fullscreen or if currently exiting, it does nothing. Then handle the "unexpected" case of having already exited in the animation frame task. (Would also match Firefox in the simple cases tested.)

Others?

The first is simple in some sense, but would enable long series of transitions in and out of fullscreen, pointless from the user's perspective. The second or third seems simpler to implement in Blink at this time.

**Comment by upsuper** [comment]
I agree with the third. We should not trigger resize if we are already in fullscreen.

The first is only simple in concept. I think that would require impls to have an additional queue for that, but it doesn't make anything better for normal usecases. I'm against adding impl complexity like that without a good reason.

**Comment by foolip** [comment]
@upsuper, can you take a look at https://github.com/whatwg/fullscreen/pull/64?

Not all error cases are handled, and even figuring out what they all are is tricky. I'll try to tinker with this in Blink, it might be useful to enumerate all the ways that things can go wrong in Gecko and how you'd like to handle them.

**Comment by foolip** [comment]
A request can fail with a fullscreenerror event, so I've also tested what events are fired for the "request then exit+request" case after clicking "exit+request" with a [variation](https://gist.github.com/foolip/c333ab539685bebc0e53d1808ce4aa21) on the original script:

<table>
<tr>
 <th>Chrome 55 on Linux</th>
 <td>fullscreenchange, fullscreenchange, fullscreenchange, resize</td>
<tr>
 <th>Edge 15 on Window 10</th>
 <td>fullscreenchange, fullscreenchange, resize, resize</td>
<tr>
 <th>Firefox Nightly (53) on Linux</th>
 <td>fullscreenchange, resize</td>
</table>

(With unprefixed API manually enabled.)

So, at least no existing browser fires a fullscreenerror event in this case. The only sensible thing for the promise is to reject it, but it would be possible to fire no events.

Opinions?

**Comment by foolip** [comment]
@annevk, do you care either way about how all of this behaves, or is anything fine as long as we get interop?

**Comment by annevk** [comment]
I don't think I care strongly, but it would be nice if what we end up with was principled somehow.

**Comment by foolip** [comment]
Yeah. I think the first option from https://github.com/whatwg/fullscreen/issues/63#issuecomment-261948581 has a simplicity of sorts, but avoiding cycles of entering and exiting would require some ugly checks that make it pointless.

Another pretty simple option is to say that globally there can only be one pending fullscreen call at a time, but that doesn't mix well with multi-process. Will continue to fiddle with tests and spec text and ask for review when I'm satisfied.
